### PR TITLE
[ENH] Add `also`

### LINKED ISSUE #731: Add an `also` method akin to Kotlins `also`
# Brief Description
I would like to propose a new function `also` that takes a function with a single dataframe for input, runs the function on the chained dataframe, and returns the input dataframe.  

I would love to take a whack at implementing this if you're interested incorporating it into the library.

As I've been trying to adopt chaining as much as possible, I've routinely bumped into scenarios where I need to break the chain for one reason or another. This API would eliminate most reasons for breaking the chain

This function is inspired by Kotlin's `also` function.  If we were to mimic Kotlin's `also`, the signature would look something like

`also: (f: T -> None) -> T` 

It would behave very similar to `pipe` or `then`, but can be used on functions that do not return anything. This can be very useful for any sort of side-effect operations within a long chain, such as periodic logging or debugging.

# Example API

Below are some examples of how this API would simplify code I've written in the past. 

### print the df while processing.
```python
# instead of defining an extra function
def print_and(df):
    print(df)
    return df

df.groupby(["column1"]).agg(sum).then(print_and).then(my_transform)

# or breaking the chain up
df_tmp = df.groupby(["column1"]).agg(sum)
print(df_tmp)
df_tmp.then(my_transform)

# you could just inline the print statement
df.groupby(["column1"]).agg(sum).also(print).then(my_transform)
```

### Insert a breakpoint while debugging, but keep going when you're done
Similar to above, to add a breakpoint for debugging, you would either have to create a new `pipe` compatible function or break the chain into smaller steps.

Instead, you could chain a breakpoint like so

```python
df.groupby(["column1"]).agg(sum).also(lambda x: breakpoint()).then(my_transform)
```

### Save a midpoint of a processing chain and then keep going

Another possibility would be running IO in the middle of the chain.  Often, a midpoint in the chain is just as important as the end result.  `also` would allow me to save the results and keep going without breaking a chain.

```
df.groupby(["column1"]).agg(sum).also(lambda x: x.to_csv("filename.csv)).then(my_transform)
```

### pipe-dream: Save a midpoint as a variable

I haven't figured out how to get this to work given Python's scoping rules, maybe the new `:=` operator will make this possible. 

This could be really helpful in scenarios where you're running groupbys or other aggregations and want to join those back into an ungrouped frame.

```
df.groupby(["column1"]).agg(sum).also(lambda x: df_mid = x).then(my_transform).assign(original_var = df_mid)
```

### PR DESCRIPTION
**Pull request by sauln**
# PR Description

Please describe the changes proposed in the pull request:

- Adds `also` method as discussed in #731 
- Adds tests for `also`
- Adds docstring for `also`

**This PR resolves #731.**

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [x] Add a line to `CHANGELOG.rst` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [x] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - flake8 checking
    - running the test suite
    - docs build

Once done, please check off the check-box above.

If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

If you are adding code changes, please ensure the following:

- [x] Ensure that you have added tests.
- [x] Run all tests (`$ pytest .`) locally on your machine.
    - [x] Check to ensure that test coverage covers the lines of code that you have added.
    - [x] Ensure that all tests pass.

## Documentation Changes

If you are adding documentation changes, please ensure the following:

- [x] Build the docs locally.
- [x] View the docs to check that it renders correctly.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by sauln** [comment]
Locally, everything passed except the tests with Spark. Also, when I build the docs there is are a ton of changes. Do you want the new files in `docs/reference` committed?

File: tests/functions/test_also.py
Comment (ericmjl) [inline tests/functions/test_also.py:2]: This is a new one for me, @sauln! Funnily enough, I've never mocked up anything in a test before. 

(Not a review, btw, just a comment.)

File: tests/functions/test_also.py
Comment (ericmjl) [inline tests/functions/test_also.py:6]:
```suggestion
def remove_first_two_letters_from_col_names(df):
    """Helper function to mutate dataframe by changing column names."""
```

File: tests/functions/test_also.py
Comment (ericmjl) [inline tests/functions/test_also.py:14]:
```suggestion
def remove_rows_3_and_4(df):
    """Helper function to mutate dataframe by removing rows."""
```

File: tests/functions/test_also.py
Comment (ericmjl) [inline tests/functions/test_also.py]:
```suggestion
def drop_inplace(df):
    """
    Helper function to mutate dataframe by dropping a column.
    
    We usually would not use `inplace=True` in a block,
    but the intent here is to test that 
    the in-place modification of a dataframe
    doesn't get passed through in the `.also()` function.
    Hence, we tell Flake8 to skip checking `PD002` on that line.
    """
    df.drop(columns=[df.columns[0]], inplace=True)  # noqa: PD002
```

Added a noqa for `PD002`, as well as the intent of doing so.

File: tests/functions/test_also.py
Comment (ericmjl) [inline tests/functions/test_also.py:35]:
```suggestion
def test_also_column_manipulation_no_change(dataframe):
    """Test that changed dataframe inside `.also()` doesn't get returned."""
```

File: tests/functions/test_also.py
Comment (ericmjl) [inline tests/functions/test_also.py:44]:
```suggestion
def test_also_remove_rows_no_change(dataframe):
    """Test that changed dataframe inside `.also()` doesn't get returned."""
```

File: tests/functions/test_also.py
Comment (ericmjl) [inline tests/functions/test_also.py:52]:
```suggestion
def test_also_runs_function(dataframe):
   """Test that `.also()` executes the function."""
```

File: tests/functions/test_also.py
Comment (ericmjl) [inline tests/functions/test_also.py:61]:
```suggestion
def test_also_args(dataframe):
    """Test that the args are passed through to the function."""
```

File: tests/functions/test_also.py
Comment (ericmjl) [inline tests/functions/test_also.py]:
```suggestion
def test_also_kwargs(dataframe):
    """Test that the kwargs are passed through to the function."""
```

File: tests/functions/test_also.py
Comment (ericmjl) [inline tests/functions/test_also.py:79]:
```suggestion
def test_also_drop_inplace(dataframe):
    """Test that in-place modification of dataframe does not pass through."""
```

File: tests/functions/test_also.py
Comment (ericmjl) [inline tests/functions/test_also.py:2]:
```suggestion
"""Unit tests for `.also()`."""
from unittest.mock import Mock
```

File: janitor/functions.py
Comment (ericmjl) [inline janitor/functions.py]:
```suggestion
    """ # noqa: E501
```

This will tell flake8 to skip the line length check on the docstring. In this particular case, I find it reasonable that line 2986 goes beyond the 79 chars that we use (I have it set to 79 because in VSCode on a 15" MBP, I can split the window left-right and still see the full source code nicely.)

**Review by ericmjl — CHANGES_REQUESTED** [review]
Very thoughtful setup on the tests, @sauln! Thanks for the effort put in 😄. I'm only requesting some changes regarding docstrings.

The `spark` tests might fail locally, but that shouldn't worry you. As long as they don't fail on the CI, which is the ultimate arbiter of whether the tests passed or not, that's cool with me.

Speaking of the CI, it was `flake8` that was failing. Some of my suggestions below should help. 

Other than that, it's just adding docstrings to the tests. I noticed I had trouble deciphering some tests that others had submitted, so I've started asking for docstrings on tests. (Hynek Schwalak has a great [blog post](https://hynek.me/articles/document-your-tests/) on this.) Please modify my suggestions if they're phrased incorrectly!

**Comment by ericmjl** [comment]
@sauln I just realized I forgot to respond to one of your questions:

> Also, when I build the docs there is are a ton of changes. Do you want the new files in docs/reference committed?

No need to worry about these files, as they are built automatically on `readthedocs` and for the backup deploy to github pages!

File: tests/functions/test_also.py
Comment (ericmjl) [inline tests/functions/test_also.py]: Oops! Just a minor nitpick - `kwargs` instead of `kargs`.

```suggestion
def test_also_kwargs(dataframe):
```

**Review by ericmjl — APPROVED** [review]
Everything else looks great. Thanks @sauln!

**Comment by sauln** [comment]
Awesome! Thank you @ericmjl for the thorough feedback and interest in this new feature :D I'm excited to make use of it in my own day-to-day.

**Comment by sauln** [comment]
Alright, it looked like I was using some features of `unittest` only available in Python 3.8.. I've reworked the tests with mocks to reflect the 3.6 API (https://docs.python.org/3.6/library/unittest.mock.html#calls-as-tuples).  

@ericmjl  it looks like `DeepSource` doesn't like the tuple unpacking the Python docs recommend. Is this something we can silence, or would you rather I rework it to avoid the `bug risk`?

**Comment by ericmjl** [comment]
@sauln hmm, I thought that the tuple unpacking inside the tests would be fine. Is there a way to rework those lines such that the bug risk is removed? If not, don't fret it, just add the `skipcq:` piece on those lines.

**Comment by ericmjl** [comment]
@sauln thanks for the changes! I’m going to invite review from at least one more core dev before we merge (or wait 2 days, whichever is earlier). This prevents the situation where only one person, i.e. myself, knows what’s going on in the codebase.

**Comment by sauln** [comment]
Sweet, thank you!

**Review by samukweku — APPROVED** [review]

**Pull request merged**
